File: win/system/backend/app/api/bot.py
"""
Discord Bot API endpoints
Secure API for bot integration
"""

from typing import List
from fastapi import APIRouter, HTTPException, Depends, status

from app.models import (
    TokenRegenerateResponse, HealthResponse,
    FronterUpdateRequest, MultiSwitchRequest,
    FronterUpdateResponse, MultiSwitchResponse
)
from app.services.bot_token_service import regenerate_bot_token
from app.services.pluralkit import get_system, get_members, get_fronters, set_front
from app.services.tag_service import enrich_members_with_tags
from app.services.status_service import enrich_members_with_status
from app.dependencies.auth import get_owner_user
from app.dependencies.bot import verify_bot_access
from app.models import User

router = APIRouter(prefix="/api/bot")

# ============================================================================
# HEALTH & TOKEN MANAGEMENT
# ============================================================================

@router.get("/health", response_model=HealthResponse)
async def bot_health_check(authenticated: bool = Depends(verify_bot_access)):
    """Health check endpoint for the bot"""
    return HealthResponse(
        status="ok",
        message="Bot API is operational",
        authenticated=authenticated
    )

@router.post("/token/regenerate", response_model=TokenRegenerateResponse)
async def regenerate_token(current_user: User = Depends(get_owner_user)):
    """Regenerate bot access token (owner only)"""
    new_token = regenerate_bot_token()
    
    return TokenRegenerateResponse(
        success=True,
        message="Bot access token has been regenerated. Update your bot's .env file.",
        new_token=new_token
    )

@router.post("/token/regenerate-self", response_model=TokenRegenerateResponse)
async def regenerate_token_self(authenticated: bool = Depends(verify_bot_access)):
    """Regenerate bot access token using current bot token"""
    new_token = regenerate_bot_token()
    
    return TokenRegenerateResponse(
        success=True,
        message="Bot access token has been regenerated. The old token is now invalid.",
        new_token=new_token
    )

# ============================================================================
# SYSTEM INFO ENDPOINTS
# ============================================================================

@router.get("/system/info")
async def get_system_info_for_bot(authenticated: bool = Depends(verify_bot_access)):
    """Get system information"""
    try:
        system_data = await get_system()
        return {"success": True, "data": system_data}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch system info: {str(e)}"
        )

@router.get("/members")
async def get_members_for_bot(authenticated: bool = Depends(verify_bot_access)):
    """Get all members with tags and status"""
    try:
        members_data = await get_members()
        members_with_tags = enrich_members_with_tags(members_data)
        members_with_status = enrich_members_with_status(members_with_tags)
        
        return {"success": True, "data": members_with_status}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch members: {str(e)}"
        )

@router.get("/fronters")
async def get_fronters_for_bot(authenticated: bool = Depends(verify_bot_access)):
    """Get current fronters"""
    try:
        fronters_data = await get_fronters()
        return {"success": True, "data": fronters_data}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch fronters: {str(e)}"
        )

# ============================================================================
# FRONTING CONTROL ENDPOINTS
# ============================================================================

@router.post("/switch", response_model=MultiSwitchResponse)
async def bot_multi_switch(
    request: MultiSwitchRequest,
    authenticated: bool = Depends(verify_bot_access)
):
    """Switch the system fronts"""
    try:
        # Validate member IDs
        all_members = await get_members()
        valid_member_ids = {member['id'] for member in all_members}
        
        invalid_ids = [mid for mid in request.member_ids if mid not in valid_member_ids]
        if invalid_ids:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid member IDs: {', '.join(invalid_ids)}"
            )
        
        # Set the front
        await set_front(request.member_ids)
        
        # Get updated fronters
        updated_fronters = await get_fronters()
        
        return MultiSwitchResponse(
            status="success",
            message="Fronters updated successfully",
            fronters=updated_fronters.get('members', []),
            count=len(request.member_ids)
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to switch fronters: {str(e)}"
        )
    
# @router

File: system/backend/src/routes/bot.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Discord Bot API endpoints.
 * Secure API for bot integration.
 *
 * Like battery.py, this router's prefix ("/api/bot") was baked into the
 * Python APIRouter itself rather than added at include_router() time, so
 * paths here are absolute and this router should be mounted directly with
 * app.use(botRouter).
 */

import { Router, type Request, type Response, type NextFunction } from 'express';

import { regenerateBotToken } from '../services/bot_token_service.js';
import { getSystem, getMembers, getFronters, setFront } from '../services/pluralkit.js';
import { enrichMembersWithTags } from '../services/tag_service.js';
import { enrichMembersWithStatus } from '../services/status_service.js';
import { requireAuth, requireOwner } from '../dependencies/auth.js';
import { verifyBotAccess } from '../dependencies/bot.js';
import { HttpError } from '../core/errors.js';
import { MultiSwitchRequestSchema } from '../models/index.js';

export const botRouter = Router();

// ============================================================================
// HEALTH & TOKEN MANAGEMENT
// ============================================================================

/** Health check endpoint for the bot */
botRouter.get('/api/bot/health', verifyBotAccess, async (_req: Request, res: Response) => {
  res.json({ status: 'ok', message: 'Bot API is operational', authenticated: true });
});

/** Regenerate bot access token (owner only) */
botRouter.post(
  '/api/bot/token/regenerate',
  requireAuth,
  requireOwner,
  async (_req: Request, res: Response) => {
    const newToken = await regenerateBotToken();

    res.json({
      success: true,
      message: "Bot access token has been regenerated. Update your bot's .env file.",
      new_token: newToken,
    });
  },
);

/** Regenerate bot access token using the current bot token */
botRouter.post(
  '/api/bot/token/regenerate-self',
  verifyBotAccess,
  async (_req: Request, res: Response) => {
    const newToken = await regenerateBotToken();

    res.json({
      success: true,
      message: 'Bot access token has been regenerated. The old token is now invalid.',
      new_token: newToken,
    });
  },
);

// ============================================================================
// SYSTEM INFO ENDPOINTS
// ============================================================================

/** Get system information */
botRouter.get(
  '/api/bot/system/info',
  verifyBotAccess,
  async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const systemData = await getSystem();
      res.json({ success: true, data: systemData });
    } catch (err) {
      next(new HttpError(500, `Failed to fetch system info: ${String(err)}`));
    }
  },
);

/** Get all members with tags and status */
botRouter.get(
  '/api/bot/members',
  verifyBotAccess,
  async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const membersData = await getMembers();
      const membersWithTags = await enrichMembersWithTags(membersData);
      const membersWithStatus = await enrichMembersWithStatus(membersWithTags);

      res.json({ success: true, data: membersWithStatus });
    } catch (err) {
      next(new HttpError(500, `Failed to fetch members: ${String(err)}`));
    }
  },
);

/** Get current fronters */
botRouter.get(
  '/api/bot/fronters',
  verifyBotAccess,
  async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const frontersData = await getFronters();
      res.json({ success: true, data: frontersData });
    } catch (err) {
      next(new HttpError(500, `Failed to fetch fronters: ${String(err)}`));
    }
  },
);

// ============================================================================
// FRONTING CONTROL ENDPOINTS
// ============================================================================

/** Switch the system fronts */
botRouter.post(
  '/api/bot/switch',
  verifyBotAccess,
  async (req: Request, res: Response, next: NextFunction) => {
    const parsed = MultiSwitchRequestSchema.safeParse(req.body);
    if (!parsed.success) {
      res.status(422).json({ detail: parsed.error.issues });
      return;
    }
    const { member_ids } = parsed.data;

    try {
      // Validate member IDs
      const allMembers = await getMembers();
      const validMemberIds = new Set(allMembers.map((m) => m.id));

      const invalidIds = member_ids.filter((mid) => !validMemberIds.has(mid));
      if (invalidIds.length > 0) {
        next(new HttpError(400, `Invalid member IDs: ${invalidIds.join(', ')}`));
        return;
      }

      // Set the front
      await setFront(member_ids);

      // Get updated fronters
      const updatedFronters = await getFronters();

      res.json({
        status: 'success',
        message: 'Fronters updated successfully',
        fronters: updatedFronters.members ?? [],
        count: member_ids.length,
      });
    } catch (err) {
      if (err instanceof HttpError) {
        next(err);
        return;
      }
      next(new HttpError(500, `Failed to switch fronters: ${String(err)}`));
    }
  },
);